File: backend/src/main/java/com/bank/ebanking/service/DepositService.java
package com.bank.ebanking.service;

import com.bank.ebanking.dto.DepositRequestDTO;
import com.bank.ebanking.entity.Account;
import com.bank.ebanking.entity.DepositRequest;
import com.bank.ebanking.entity.Transaction;
import com.bank.ebanking.repository.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
public class DepositService {

    private final DepositRequestRepository depositRequestRepository;
    private final AccountRepository        accountRepository;
    private final TransactionRepository    transactionRepository;
    private final UserRepository           userRepository;
    private final AuditLogService          auditLogService;

    public DepositService(DepositRequestRepository depositRequestRepository,
                          AccountRepository accountRepository,
                          TransactionRepository transactionRepository,
                          UserRepository userRepository,
                          AuditLogService auditLogService) {
        this.depositRequestRepository = depositRequestRepository;
        this.accountRepository        = accountRepository;
        this.transactionRepository    = transactionRepository;
        this.userRepository           = userRepository;
        this.auditLogService          = auditLogService;
    }

    // ─── CLIENT soumet une demande de dépôt ──────────────
    @Transactional
    public DepositRequestDTO createRequest(Long accountId,
                                           java.math.BigDecimal amount,
                                           String description,
                                           String clientUsername) {
        @SuppressWarnings("null")
        Account account = accountRepository.findById(accountId)
                .orElseThrow(() -> new RuntimeException("Account not found"));

        if (!"ACTIVE".equals(account.getStatus())) {
            throw new RuntimeException("Account is not active");
        }
        if (amount == null || amount.compareTo(java.math.BigDecimal.ZERO) <= 0) {
            throw new RuntimeException("Amount must be greater than zero");
        }

        DepositRequest request = new DepositRequest();
        request.setAccount(account);
        request.setAmount(amount);
        request.setDescription(description);
        request.setStatus("PENDING");

        depositRequestRepository.save(request);

        auditLogService.log(clientUsername, "DEPOSIT_REQUEST",
                "Client requested deposit of " + amount +
                " for account " + account.getAccountNumber());

        return toDTO(request);
    }

    // ─── TELLER approuve une demande ─────────────────────
    @SuppressWarnings("null")
    @Transactional
    public DepositRequestDTO approveRequest(Long requestId,
                                             String tellerUsername) {
        DepositRequest request = depositRequestRepository.findById(requestId)
                .orElseThrow(() -> new RuntimeException("Request not found"));

        if (!"PENDING".equals(request.getStatus())) {
            throw new RuntimeException("Request already processed");
        }

        Account account = request.getAccount();

        // Créditer le compte
        account.setBalance(account.getBalance().add(request.getAmount()));
        accountRepository.save(account);

        // Créer la transaction
        Transaction tx = new Transaction();
        tx.setTransId(UUID.randomUUID().toString());
        tx.setAccount(account);
        tx.setAmount(request.getAmount());
        tx.setType("DEPOSIT");
        tx.setDescription(request.getDescription() != null
                ? request.getDescription()
                : "Deposit approved by teller");
        tx.setStatus("COMPLETED");
        transactionRepository.save(tx);

        // Mettre à jour la demande
        request.setStatus("APPROVED");
        request.setProcessedAt(LocalDateTime.now());
        userRepository.findByUsername(tellerUsername)
                .ifPresent(request::setProcessedBy);
        depositRequestRepository.save(request);

        auditLogService.log(tellerUsername, "DEPOSIT_APPROVED",
                "Teller approved deposit of " + request.getAmount() +
                " for account " + account.getAccountNumber());

        return toDTO(request);
    }

    // ─── TELLER rejette une demande ───────────────────────
    @SuppressWarnings("null")
    @Transactional
    public DepositRequestDTO rejectRequest(Long requestId,
                                            String tellerUsername) {
        DepositRequest request = depositRequestRepository.findById(requestId)
                .orElseThrow(() -> new RuntimeException("Request not found"));

        if (!"PENDING".equals(request.getStatus())) {
            throw new RuntimeException("Request already processed");
        }

        request.setStatus("REJECTED");
        request.setProcessedAt(LocalDateTime.now());
        userRepository.findByUsername(tellerUsername)
                .ifPresent(request::setProcessedBy);
        depositRequestRepository.save(request);

        auditLogService.log(tellerUsername, "DEPOSIT_REJECTED",
                "Teller rejected deposit of " + request.getAmount() +
                " for account " + request.getAccount().getAccountNumber());

        return toDTO(request);
    }

    // ─── Liste toutes les demandes PENDING (pour TELLER) ──
    public List<DepositRequestDTO> getPendingRequests() {
        return depositRequestRepository
                .findByStatusOrderByCreatedAtAsc("PENDING")
                .stream().map(this::toDTO).collect(Collectors.toList());
    }

    // ─── Historique des demandes d'un compte (CLIENT) ─────
    public List<DepositRequestDTO> getRequestsByAccount(Long accountId) {
        return depositRequestRepository
                .findByAccountIdOrderByCreatedAtDesc(accountId)
                .stream().map(this::toDTO).collect(Collectors.toList());
    }

    // ─── Helper ───────────────────────────────────────────
    private DepositRequestDTO toDTO(DepositRequest r) {
        DepositRequestDTO dto = new DepositRequestDTO();
        dto.setId(r.getId());
        dto.setAccountId(r.getAccount().getId());
        dto.setAccountNumber(r.getAccount().getAccountNumber());
        dto.setAmount(r.getAmount());
        dto.setDescription(r.getDescription());
        dto.setStatus(r.getStatus());
        dto.setCreatedAt(r.getCreatedAt());
        dto.setProcessedAt(r.getProcessedAt());

        // Nom du client
        if (r.getAccount().getUser() != null) {
            dto.setClientName(
                r.getAccount().getUser().getFirstName() + " " +
                r.getAccount().getUser().getLastName()
            );
        }

        if (r.getProcessedBy() != null) {
            dto.setProcessedBy(r.getProcessedBy().getUsername());
        }

        return dto;
    }
}

File: ebanking_app/lib/services/deposit_service.dart
import 'dart:convert';
import '../core/constants.dart';
import '../core/api_client.dart';
import '../../models/deposit_request.dart';

class DepositService {

  static const String _baseUrl = '${AppConstants.baseUrl}/api/deposits';

  // ── CLIENT soumet une demande ─────────────────────────
  Future<DepositRequestModel?> requestDeposit(
      int accountId, double amount, {String? description}) async {
    try {
      final params = {
        'amount': amount.toString(),
        if (description != null && description.isNotEmpty)
          'description': description,
      };
      final response = await ApiClient.postWithParams(
          '$_baseUrl/request/$accountId', params);
      if (response.statusCode == 200) {
        return DepositRequestModel.fromJson(jsonDecode(response.body));
      }
      return null;
    } catch (_) { return null; }
  }

  // ── TELLER : liste des demandes PENDING ───────────────
  Future<List<DepositRequestModel>> getPendingRequests() async {
    try {
      final response = await ApiClient.get('$_baseUrl/pending');
      if (response.statusCode == 200) {
        final List data = jsonDecode(response.body);
        return data.map((j) => DepositRequestModel.fromJson(j)).toList();
      }
      return [];
    } catch (_) { return []; }
  }

  // ── TELLER : approuver ────────────────────────────────
  Future<bool> approveRequest(int requestId) async {
    try {
      final response = await ApiClient.post(
          '$_baseUrl/approve/$requestId');
      return response.statusCode == 200;
    } catch (_) { return false; }
  }

  // ── TELLER : rejeter ──────────────────────────────────
  Future<bool> rejectRequest(int requestId) async {
    try {
      final response = await ApiClient.post(
          '$_baseUrl/reject/$requestId');
      return response.statusCode == 200;
    } catch (_) { return false; }
  }

  // ── Historique demandes d'un compte ───────────────────
  Future<List<DepositRequestModel>> getHistory(int accountId) async {
    try {
      final response = await ApiClient.get('$_baseUrl/history/$accountId');
      if (response.statusCode == 200) {
        final List data = jsonDecode(response.body);
        return data.map((j) => DepositRequestModel.fromJson(j)).toList();
      }
      return [];
    } catch (_) { return []; }
  }
}